v3.6.12: `hooks metrics` shows zeros + TypeError(null.toFixed) despite `hooks post-task` confirming outcomePersisted; metrics JSONs not updated by post-* writes

## Summary

In Ruflo `v3.6.12`, the `hooks metrics` dashboard (both via CLI `npx ruflo hooks metrics` and via the MCP tool `mcp__ruflo__hooks_metrics`) reports **all zeros** and crashes with `TypeError: Cannot read properties of null (reading 'toFixed')` even when `hooks post-task`, `hooks post-edit`, and `hooks post-bash` calls return success and confirm `outcomePersisted: true` with real `trajectoryId` values.

Independent investigation shows the cause: **`hooks post-*` writes go to `.swarm/memory.db` (sql.js + HNSW backend) but the dashboard reads from `.claude-flow/metrics/*.json`, and the two are not connected.** The metric JSON files are last-modified at project init time and never updated by `hooks post-*` calls.

Additionally, `hooks pretrain` reports `Patterns Extracted: 0` and `Duration: NaNs` after analyzing 51,611 files, indicating a separate computation/persistence bug in the pretrain pipeline.

This appears related to (but distinct from) closed issues #1145 and #1132 which fixed similar `toFixed` crashes in `hooks intelligence stats` / `hooks intelligence init`. Open issue #1074 is in the same family. The maintainers have actively patched this category before, so filing this is in keeping with that pattern.

---

## Environment

| Component | Version |
| --- | --- |
| `ruflo` (npm) | 3.6.12 (latest) |
| `claude-flow` (npm, alias of ruflo) | 3.6.12 |
| `@claude-flow/cli` | 3.6.12 |
| Node.js | v22.13.1 |
| npm | 10.9.2 |
| Claude Code | 2.1.126 |
| OS | Darwin 25.4.0 arm64 (macOS, Apple Silicon) |
| Project init age | 25 days (initialized 2026-04-05) |
| Init path used | `npx ruflo@latest init upgrade --settings` (post-`claude-flow init` migration) |

---

## What we did (full sequence)

1. Project was originally initialized months ago with `claude-flow init`, before the rename to Ruflo. The legacy `.claude/helpers/hook-handler.cjs` etc. shipped with that init. Confirmed by the file header comment `Claude Flow Hook Handler (Cross-Platform)`.

2. To migrate the project to current Ruflo, we ran:
   ```bash
   cd <project-root>
   cp .claude/settings.json .claude/settings.json.pre-ruflo-upgrade.bak
   cp -r .claude/helpers .claude/helpers.pre-ruflo-upgrade.bak
   npx ruflo@latest init upgrade --settings --verbose
   ```
   Output reported successful merge of helpers + settings:
   ```
   ✓ .claude/helpers/auto-memory-hook.mjs
   ✓ .claude/helpers/hook-handler.cjs
   ✓ .claude/helpers/intelligence.cjs
   ✓ .claude/helpers/statusline.cjs
   ✓ .claude/settings.json
   ```

3. `npx ruflo init check` passed: `[OK] RuFlo is initialized`.

4. `npx ruflo doctor` passed 10/14, daemon flagged as not running. We started it:
   ```bash
   npx ruflo daemon start
   ```
   Daemon came up cleanly, PID 48771, 5 active workers (`map`, `audit`, `optimize`, `consolidate`, `testgaps`).

5. We verified the MCP `hooks_post-task` write path works end-to-end:
   ```
   mcp__ruflo__hooks_post-task → response:
   {
     "taskId": "verify-pipeline-2026-04-30-2245",
     "success": true,
     "duration": 23,
     "learningUpdates": {
       "patternsUpdated": 1,
       "newPatterns": 1,
       "trajectoryId": "traj-1777603418863",
       "controller": "bridge-store",
       "outcomePersisted": true
     },
     "quality": 0.9,
     "feedback": { "recorded": true, "controller": "bridge-store", "updates": 1 }
   }
   ```
   And the `mcp__ruflo__memory_stats` `totalEntries` count grew (from 6 → 10 → 13 → 18 → 33 across multiple write calls). So writes ARE landing in `.swarm/memory.db`.

---

## Bug 1 — `hooks metrics` shows all zeros despite confirmed writes

After multiple successful `hooks post-task`, `hooks post-edit`, and `hooks post-bash` calls (each returning `success: true` / `outcomePersisted: true`):

```bash
$ npx ruflo hooks metrics
Learning Metrics Dashboard (24h)

📊 Pattern Learning
+----------------+-------+
| Metric         | Value |
+----------------+-------+
| Total Patterns |     0 |
| Successful     |     0 |
| Failed         |     0 |
| Avg Confidence |  0.0% |
+----------------+-------+

🤖 Agent Routing
+------------------+-------+
| Metric           | Value |
+------------------+-------+
| Routing Accuracy |  0.0% |
| Total Routes     |     0 |
| Top Agent        |       |
+------------------+-------+

⚡ Command Execution
[ERROR] Unexpected error: TypeError: Cannot read properties of null (reading 'toFixed')
```

Same result via the MCP tool:
```js
mcp__ruflo__hooks_metrics period:1h →
{
  "_real": true,
  "_note": "No metrics data collected yet. Data populates from hooks_post-task, hooks_post-edit, hooks_post-command, and hooks_route calls.",
  "patterns": { "total": 0, "successful": 0, "failed": 0, "avgConfidence": null },
  "agents": { "routingAccuracy": null, "totalRoutes": 0, "topAgent": null },
  "commands": { "totalExecuted": 0, "successRate": null, "avgRiskScore": null }
}
```

The `_note` says metrics SHOULD populate from `hooks_post-task` calls — but they don't.

### Investigation — why dashboard sees zero

`md5sum` of `.claude-flow/metrics/*.json` BEFORE and AFTER calling `npx ruflo hooks post-task -t "test" --success`:

```
BEFORE:
de1f6dc54d10d158bc21e03720079f04  codebase-map.json
a8010ba278d4a671d4336c6d11a8684a  learning.json
d850dfbd168c750163bd57e36ab43d3c  swarm-activity.json
559afd096a65ef98d3f30336d0c3ef0c  v3-progress.json

POST-TASK CALL CONFIRMED: "Patterns Updated: 1, New Patterns: 1, Trajectory ID: traj-1777603489364"

AFTER:
de1f6dc54d10d158bc21e03720079f04  codebase-map.json    ← unchanged
a8010ba278d4a671d4336c6d11a8684a  learning.json        ← unchanged
d850dfbd168c750163bd57e36ab43d3c  swarm-activity.json  ← unchanged
559afd096a65ef98d3f30336d0c3ef0c  v3-progress.json     ← unchanged
```

`learning.json` last-modified `Apr 5 01:16` — 25 days ago, project init time, never touched since. Yet `hooks post-task` writes succeed (per response payload) and `memory_stats` confirms entries are growing in the sql.js + HNSW backend.

Conclusion: **`hooks post-*` writes to `.swarm/memory.db` but the dashboard reads from `.claude-flow/metrics/*.json`, and there is no bridge between the two stores.** Either the dashboard should read from `memory.db`, or `hooks post-*` should also update the JSON files, or a daemon worker should bridge them on a schedule. None of these is happening.

### Root cause of the `TypeError: Cannot read properties of null (reading 'toFixed')`

`learning.json` contains:
```json
{
  "routing": { "accuracy": 0, "decisions": 0 },
  "patterns": { "shortTerm": 0, "longTerm": 0, "quality": 0 },
  "sessions": { "total": 0, "current": null }
}
```

The "Command Execution" section of the metrics dashboard appears to call `.toFixed()` on a value derived from `sessions.current` (which is `null`) without a null-check. This is the same defensive-programming gap pattern fixed in #1145 and #1132 for `hooks intelligence stats` / `hooks intelligence init`, but it persists in `hooks metrics`.

---

## Bug 2 — `hooks pretrain` reports `Patterns Extracted: 0` and `Duration: NaNs` after analyzing 51,611 files

```bash
$ npx ruflo hooks pretrain
Pretraining Intelligence (4-Step Pipeline + Embeddings)

[ pipeline runs through RETRIEVE → JUDGE → DISTILL → CONSOLIDATE → EMBED → HYPERBOLIC ]

Pretraining completed
+-------------------------+-------+
| Metric                  | Value |
+-------------------------+-------+
| Files Analyzed          | 51611 |
| Patterns Extracted      |     0 |
| Strategies Learned      |       |
| Trajectories Evaluated  |       |
| Contradictions Resolved |       |
| Duration                |  NaNs |
+-------------------------+-------+

[OK] Repository intelligence bootstrapped successfully
```

After this completes:
- `Patterns Extracted: 0` — analyzing 51,611 files yielded zero patterns
- `Strategies Learned: (empty)` — display gap (likely undefined value)
- `Trajectories Evaluated: (empty)` — display gap
- `Contradictions Resolved: (empty)` — display gap
- `Duration: NaNs` — JavaScript NaN being stringified into the table

Re-running `hooks metrics` after pretrain returns the same zeros + same `toFixed` crash, so pretrain doesn't populate the dashboard either.

---

## Reproduction (clean steps, copy-paste safe)

```bash
# 1. Fresh project init
mkdir /tmp/ruflo-repro && cd /tmp/ruflo-repro
git init
npx ruflo@latest init --minimal

# 2. Confirm metrics dashboard is initially empty
npx ruflo hooks metrics    # expect: all zeros, no crash (yet)

# 3. Record a task outcome
npx ruflo hooks post-task -t "test" --success --quality 0.9
# expect output: "Patterns Updated: 1, New Patterns: 1, Trajectory ID: traj-..."

# 4. Re-check metrics — STILL all zeros, AND now crashes:
npx ruflo hooks metrics
# expect: "TypeError: Cannot read properties of null (reading 'toFixed')"
#         in the "⚡ Command Execution" section

# 5. Confirm the underlying memory.db DOES have the entry
sqlite3 .swarm/memory.db "SELECT COUNT(*) FROM memory_entries"
# expect: > 0 (writes are landing)

# 6. Confirm metric JSONs are stale
md5sum .claude-flow/metrics/*.json   # capture
npx ruflo hooks post-task -t "another" --success
md5sum .claude-flow/metrics/*.json   # identical to before — JSONs not written

# 7. Try pretrain
npx ruflo hooks pretrain
# expect: "Patterns Extracted: 0", "Duration: NaNs", empty Strategies/Trajectories/Contradictions
```

---

## Expected behavior

1. After `hooks post-task` returns `outcomePersisted: true`, `hooks metrics` should show that pattern in the `Total Patterns` count within the appropriate period window (or at least eventually, after a documented bridge interval).
2. The "Command Execution" section should not crash when `learning.json.sessions.current` is `null` — needs a null-check / default value (same as the fixes for #1145 and #1132 in adjacent commands).
3. `hooks pretrain` should either extract real patterns from analyzed files or report a meaningful error (not silently emit `0` + `NaN`).
4. The relationship between `.swarm/memory.db` and `.claude-flow/metrics/*.json` should be documented OR they should be unified into a single source of truth.

---

## Workaround we adopted

Since `hooks post-*` writes DO land in `.swarm/memory.db` (verified via `mcp__ruflo__memory_stats`), we wired the project's Claude Code hooks (`PostToolUse`, `UserPromptSubmit`, `Stop`) to call `npx ruflo hooks post-edit/post-bash/route/post-task` on every event by adding parallel hook commands to the project `.claude/settings.json`. This ensures memory is captured per session even though the dashboard view of that data is broken.

We are NOT relying on the metrics dashboard for production observability — only on `mcp__ruflo__memory_stats` and direct `sqlite3 .swarm/memory.db` queries.

---

## Related issues

- #1145 (closed Mar 17, 2026): `hooks intelligence stats` `.toFixed()` of undefined → fixed
- #1132 (closed Mar 19, 2026): `hooks intelligence init` `Cannot read properties of undefined (reading 'toFixed')` → fixed
- #1074 (open since Feb 4, 2026): `hooks intelligence --status: TypeError on SONA display when no training data exists` → still open

The `hooks metrics` view appears to have the same defensive-programming gap that was patched in #1145 and #1132 for adjacent commands. Likely a missed code path during those PRs.

---

## Suggested fix areas

1. **`hooks metrics` Command Execution section**: add `if (!current || typeof current.successRate !== 'number') return ...` guard before `.toFixed()` calls (mirror the patches in #1145/#1132).
2. **Bridge between memory.db and metrics JSONs**: either point the dashboard at `.swarm/memory.db` (single source of truth), OR have `hooks post-task/post-edit/post-bash/route` ALSO update `learning.json` / `swarm-activity.json` synchronously, OR have the daemon's `consolidate` worker do the bridging on a schedule and document the lag.
3. **`hooks pretrain` Patterns Extracted: 0 + Duration: NaNs**: the pipeline is producing data (`Files Analyzed: 51611` is non-zero) but not persisting/computing the pattern count or duration. Likely a Promise-resolution or accumulator-init bug in the pipeline.

---

*Filed by Stuart Kerr (@stuinfla) with diagnostic assistance from Claude Code. Diagnostic transcript and md5 snapshots available on request — every claim above traces to a tool-call probe captured in the session log.*

--- a/v3/@claude-flow/cli/src/commands/hooks.ts
+++ b/v3/@claude-flow/cli/src/commands/hooks.ts
@@ -11,6 +11,17 @@ import { storeCommand } from './transfer-store.js';
 import { existsSync, readFileSync, statSync } from 'node:fs';
 import { join } from 'node:path';
 
+/**
+ * #1686 — `?? 0` only defaults null/undefined; NaN slips through and
+ * surfaces as `"NaN"` (or earlier crashed `.toFixed`) in the metrics
+ * dashboard and pretrain output. Coerce to a finite number, fall back
+ * to `fallback` when the input is null/undefined/non-numeric/NaN/Infinity.
+ */
+function safeNum(value: unknown, fallback = 0): number {
+  const n = typeof value === 'number' ? value : Number(value);
+  return Number.isFinite(n) ? n : fallback;
+}
+
 // ============================================================================
 // Coverage Data Reader - reads Jest/Istanbul coverage files from disk
 // ============================================================================
@@ -1113,8 +1124,10 @@ const pretrainCommand: Command = {
       spinner.succeed('Pretraining completed');
 
       // Normalize shape: prefer `statistics`, fall back to `stats` for older tools.
+      // #1686 — coerce duration through safeNum so a NaN from the underlying
+      // pretrain pipeline surfaces as `0.0s` rather than `NaNs`.
       const stats = (rawResult.statistics ?? rawResult.stats ?? {}) as Record<string, number | undefined>;
-      const durationMs = rawResult.duration ?? (rawResult.statistics?.executionTime as number | undefined) ?? 0;
+      const durationMs = safeNum(rawResult.duration ?? rawResult.statistics?.executionTime);
       const result = { ...rawResult, stats, duration: durationMs };
 
       if (ctx.flags.format === 'json') {
@@ -1375,20 +1388,21 @@ const metricsCommand: Command = {
       });
 
       // Normalize across both shapes; default every numeric to 0 so toFixed
-      // never sees null/undefined.
-      const totalPatterns = rawMetrics.patterns?.total ?? rawMetrics.summary?.patternsLearned ?? 0;
-      const successfulPatterns = rawMetrics.patterns?.successful ?? Math.round((rawMetrics.summary?.successRate ?? 0) * totalPatterns);
-      const failedPatterns = rawMetrics.patterns?.failed ?? Math.max(0, totalPatterns - successfulPatterns);
-      const avgConfidence = rawMetrics.patterns?.avgConfidence ?? rawMetrics.summary?.avgQuality ?? 0;
+      // never sees null/undefined. #1686 — also coerce NaN through `safeNum`
+      // because `?? 0` only catches null/undefined; an upstream NaN would
+      // still land in `.toFixed(...)` and surface as `"NaN"`.
+      const totalPatterns = safeNum(rawMetrics.patterns?.total ?? rawMetrics.summary?.patternsLearned);
+      const successfulPatterns = safeNum(rawMetrics.patterns?.successful ?? Math.round(safeNum(rawMetrics.summary?.successRate) * totalPatterns));
+      const failedPatterns = Math.max(0, safeNum(rawMetrics.patterns?.failed ?? totalPatterns - successfulPatterns));
+      const avgConfidence = safeNum(rawMetrics.patterns?.avgConfidence ?? rawMetrics.summary?.avgQuality);
 
-      const routingAccuracy = rawMetrics.agents?.routingAccuracy
-        ?? (rawMetrics.routing?.avgConfidence ?? 0);
-      const totalRoutes = rawMetrics.agents?.totalRoutes ?? rawMetrics.routing?.totalRoutes ?? 0;
+      const routingAccuracy = safeNum(rawMetrics.agents?.routingAccuracy ?? rawMetrics.routing?.avgConfidence);
+      const totalRoutes = safeNum(rawMetrics.agents?.totalRoutes ?? rawMetrics.routing?.totalRoutes);
       const topAgent = rawMetrics.agents?.topAgent ?? rawMetrics.routing?.topAgents?.[0]?.agent ?? 'n/a';
 
-      const totalCommands = rawMetrics.commands?.totalExecuted ?? rawMetrics.commands?.totalCommands ?? 0;
-      const commandSuccessRate = rawMetrics.commands?.successRate ?? 0;
-      const avgRiskScore = rawMetrics.commands?.avgRiskScore ?? rawMetrics.commands?.avgExecutionTime ?? 0;
+      const totalCommands = safeNum(rawMetrics.commands?.totalExecuted ?? rawMetrics.commands?.totalCommands);
+      const commandSuccessRate = safeNum(rawMetrics.commands?.successRate);
+      const avgRiskScore = safeNum(rawMetrics.commands?.avgRiskScore ?? rawMetrics.commands?.avgExecutionTime);
 
       const result = {
         ...rawMetrics,

